### PR TITLE
Split normalizers and denormalizers in Serializer

### PR DESCRIPTION
**Pull request by 243083df**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Comment by 243083df** [comment]
@nicolas-grekas what to do with exceptions and does we need to keep index from incoming array, because current realization only keep priority?

**Comment by nicolas-grekas** [comment]
I've no idea what you're trying to achieve and why so I cannot comment sorry...
Usually, failing test cases are the symptom of a BC break, so you'll need to think about what they mean in this regard.

**Comment by 243083df** [comment]
Yes it break BC on exception class level. But exception now more correct. Second it has minor speedup of cache warmup. Third since denormalizer and normalizer has separate cache, i think separate registered normalizers and denormalizers unify internal code

**Review by dunglas — APPROVED** [review]
I like it, but the change on the cache must be double-checked and profile to be sure there are no performance regressions.

**Comment by 243083df** [comment]
How can should i do it?

File: src/Symfony/Component/Serializer/Serializer.php [hunk 1]
```diff
@@ -83,12 +86,17 @@ public function __construct(array $normalizers = [], array $encoders = [])
...
+            }
+            if ($normalizer instanceof DenormalizerInterface) {
+                $this->denormalizers[] = $normalizer;
+            }
             if (!($normalizer instanceof NormalizerInterface || $normalizer instanceof DenormalizerInterface)) {
                 @trigger_error(sprintf('Passing normalizers ("%s") which do not implement either "%s" or "%s" has been deprecated since Symfony 4.2.', \get_class($normalizer), NormalizerInterface::class, DenormalizerInterface::class), E_USER_DEPRECATED);
                 // throw new \InvalidArgumentException(\sprintf('The class "%s" does not implement "%s" or "%s".', \get_class($normalizer), NormalizerInterface::class, DenormalizerInterface::class));
             }
         }
-        $this->normalizers = $normalizers;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Serializer.php]: I think we need to keep all the code infrastructure related to this property to keep BC. Then we'll be able to remove the code in 5.0.

File: src/Symfony/Component/Serializer/Serializer.php [hunk 2]
```diff
@@ -59,8 +59,11 @@ class Serializer implements SerializerInterface, ContextAwareNormalizerInterface
      * @internal since Symfony 4.1
      */
     protected $normalizers = [];
+    /**
+     * @internal since Symfony 4.4
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Serializer.php]: let's make the property private then

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]

**Comment by fabpot** [comment]
Closing as there is no more activities. Feel free to reopen if you have time to finish the work.

**Pull request closed without merging**